src/serialization/deserializer.rs

```diff
@@ -57,10 +58,29 @@ impl Deserializer {
                                             -> Result<Histogram<T>, DeserializeError> {
         let cookie = reader.read_u32::<BigEndian>()?;
 
-        if cookie != V2_COOKIE {
+        return match cookie {
+            V2_COOKIE => self.deser_v2(reader),
+            V2_COMPRESSED_COOKIE => self.deser_v2_compressed(reader),
+            _ => Err(DeserializeError::InvalidCookie)
+        }
+    }
+
+    fn deser_v2_compressed<T: Counter, R: Read>(&mut self, reader: &mut R) -> Result<Histogram<T>, DeserializeError> {
+        let payload_len = reader.read_u32::<BigEndian>()?.to_usize()
+            .ok_or(DeserializeError::UsizeTypeTooSmall)?;
+
+        // TODO reuse deflate buf, or switch to lower-level flate2::Decompress
+        let mut deflate_reader = DeflateDecoder::new(reader.take(payload_len as u64));
```

Comment:
Yeah, but mostly for intellectual vanity reasons (and a little bit because I want to keep the way to no-heap open) I want serialization and deserialization to be allocation-free. :)